### PR TITLE
add key-value file!

### PR DESCRIPTION
**Pull request by ruclz**
Oceanbase sort needs key-value struct. But the avx512-64bits-qsort only supports key sort. We need key-value sort in this lib.

### CODE REVIEW

**Comment by xtangxtang** [comment]
The PR is to meet the requirement from Oceanbase.

Oceanbase is a Key/Value store Database. It uses std::sort to sort the object during creating index or other jobs. Not like just sort the key,  Oceanbase passes inner ObStore class object as an array to do the sort job. And the std::sort callback comparing function to determine how to get the key from the ObStore to compare. So in Oceanbase, it swaps the ObStore object index in the array according to the sort function.

Since the avx512 sort in this project do a wondful work in bitonic sort, we want to reuse the core code. But the sort API just do the sort work, doesn't return the indexes of the key after sorted, it is hard for us to integrated the code into Oceanbase. So we create the PR to modify the sort API from _**void avx512_qsort(uint16_t *arr, int64_t arrsize)**_ to _**void avx512_qsort(uint16_t *arr,uint64_t *indexes, int64_t arrsize')**_.  The returned **_indexes_** is used to record the sorted key indexes and help API caller to swap the its objec array outside.

Currently we have created a PR in Oceanbase to merge this project code as a 3rd party lib, now is wating for this PR approval.

File: .vscode/settings.json [hunk 1]
```diff
@@ -0,0 +1,57 @@
+{
```
Comment (r-devulap) [inline .vscode/settings.json:1]: Please remove this file.

File: Makefile [hunk 2]
```diff
@@ -15,10 +15,10 @@ LD_FLAGS	= -L /usr/local/lib -l $(GTEST_LIB) -l pthread
 all : test bench
 
 $(TESTDIR)/%.o : $(TESTDIR)/%.cpp $(SRCS)
-		$(CXX) -march=icelake-client -O3 $(CXXFLAGS) -c $< -o $@
+		$(CXX) -march=icelake-client -O3 $(CXXFLAGS) -l $(GTEST_LIB) -c $< -o $@
```
Comment (r-devulap) [inline Makefile:18]: Why is this needed? This is only building object files and doesn't need any linking to done.

File: Makefile [hunk 3]
```diff
@@ -15,10 +15,10 @@ LD_FLAGS	= -L /usr/local/lib -l $(GTEST_LIB) -l pthread
...
 
 test: $(TESTDIR)/main.cpp $(TESTOBJS) $(SRCS)
-		$(CXX) tests/main.cpp $(TESTOBJS) $(CXXFLAGS) $(LD_FLAGS) -o testexe
+		$(CXX) tests/main.cpp $(TESTOBJS) $(CXXFLAGS) $(LD_FLAGS) -o testexe
```
Comment (r-devulap) [inline Makefile:21]: Unnecessary whitespace.

**Review by r-devulap — CHANGES_REQUESTED** [review]
As far as I understood, you are essentially adding argsort. The indices you pass is also sorted along with the array. Since this is a new feature, could you add new unit tests that cover this feature? 

In general, the PR needs a lot of formatting changes. The diff is pretty hard to read. Could you fix the formatting and run clang-format on the files and re-submit? The formatting specifications are in a _clang-format file.

**Comment by r-devulap** [comment]
On a second thought, argsort seems like a special case. You are sorting both key and value arrays.

**Comment by ruclz** [comment]
> As far as I understood, you are essentially adding argsort. The modified indices would sort the original array. Correct? Since this is a new feature, could you add new unit tests that cover this feature?
> 
> In general, the PR needs a lot of formatting changes. The diff is pretty hard to read. Could you fix the formatting and run clang-format on the files and re-submit? The formatting specifications are in a _clang-format file.

image

Hello, I run the clang_format execution in the project, but I got the error like the figure, can you have some suggests to solve it? Also, the unit tests can't run on my environment including the unit tests  that you have completed, can you share  your environment to me, like OS, compiler version and so on.

**Comment by ruclz** [comment]
> On a second thought, argsort seems like a special case. You are sorting both key and value arrays.

OK, we are sorting key, and the value column exchange with each other according key.

**Comment by r-devulap** [comment]
> Hello, I run the clang_format execution in the project, but I got the error like the figure, can you have some suggests to solve it? 

Hmm, not sure why you see that error. I am able to run clang-format on your patch just fine. 

> Also, the unit tests can't run on my environment including the unit tests that you have completed, can you share your environment to me, like OS, compiler version and so on.

Do you see an error? I am using ubuntu latest, g++-10 and running on a SKX. If your processor doesnt support AVX-512, you could run on an [SDE ](https://www.intel.com/content/www/us/en/developer/articles/tool/software-development-emulator.html) too with this command: `sde-external-9.14.0-2022-10-25-lin/sde64 -icx -- ./testexe`

**Comment by ruclz** [comment]
OK, I added a unittest for key-value sort, and I run clang_format on all files that I modified. Please help to review. Thanks.

**Comment by r-devulap** [comment]
Hi @ruclz This PR modifies the existing `avx512_qsort<T>()` function signature by adding another argument to the function and also adds unnecessary code if someone wants to use only data sorting and not key-value sort. I think a better option would be to move your code to new header files `avx512_64bit_keyvaluesort.hpp` and `avx512-common-keyvaluesort.h `. I understand there might be some code duplication but we can work on minimizing that in a separate PR.

**Review by r-devulap — CHANGES_REQUESTED** [review]
Move code to separate files and add new API's `avx512_keyvaluesort<T>(`).

**Comment by ruclz** [comment]
Hi, @r-devulap, I have moved my code to new header files avx512-64bit-keyvaluesort.hpp and avx512-common-keyvaluesort.h. The unit test can run correctly, please help to review, thanks.

**Comment by ruclz** [comment]
Sorry, I forgot to merge the test_all.cpp in the last commit.

File: src/avx512-64bit-keyvaluesort.hpp [hunk 4]
```diff
@@ -0,0 +1,1276 @@
...
+ * https://en.wikipedia.org/wiki/Bitonic_sorter#/media/File:BitonicSort.svg)
+ */
+// ZMM                  7, 6, 5, 4, 3, 2, 1, 0
+#define NETWORK_64BIT_1 4, 5, 6, 7, 0, 1, 2, 3
+#define NETWORK_64BIT_2 0, 1, 2, 3, 4, 5, 6, 7
+#define NETWORK_64BIT_3 5, 4, 7, 6, 1, 0, 3, 2
+#define NETWORK_64BIT_4 3, 2, 1, 0, 7, 6, 5, 4
+
+template <>
+struct zmm_kv_vector<int64_t> {
```
Comment (r-devulap) [inline src/avx512-64bit-keyvaluesort.hpp:24]: Is there a difference between `zmm_kv_vector `struct and `zmm_vector `struct in `avx512-64bit-qsort.hpp`? If not, we should move all these to a new header file `avx512-64bit-common.h` and not duplicate them. You could include the NETWORK_64BIT_* constants too.

File: src/avx512-64bit-keyvaluesort.hpp [hunk 5]
```diff
@@ -0,0 +1,1276 @@
...
+                         uint64_t *indexes,
+                         int64_t left,
+                         int64_t right,
+                         int64_t max_iters)
+{
+    /*
+     * Resort to std::sort if quicksort isnt making any progress
+     */
+    if (max_iters <= 0) {
+        heap_sort<vtype>(keys + left, indexes + left, right - left + 1);
```
Comment (r-devulap) [inline src/avx512-64bit-keyvaluesort.hpp:1194]: why not use std::sort? Just like how you used it in test_all.cpp.

File: src/avx512-64bit-keyvaluesort.hpp [hunk 6]
```diff
@@ -0,0 +1,1276 @@
...
+    int64_t pivot_index = partition_avx512<vtype>(
+            keys, indexes, left, right + 1, pivot, &smallest, &biggest);
+    if (pivot != smallest)
+        qsort_64bit_<vtype>(
+                keys, indexes, left, pivot_index - 1, max_iters - 1);
+    if (pivot != biggest)
+        qsort_64bit_<vtype>(keys, indexes, pivot_index, right, max_iters - 1);
+}
+
+X86_SIMD_SORT_INLINE int64_t replace_nan_with_inf_kv(double *arr,
```
Comment (r-devulap) [inline src/avx512-64bit-keyvaluesort.hpp:1219]: This is the same function as in `avx512-64bit-qsort.hpp`. You could move this to the common header file as well and no need to re-write it.

File: src/avx512-64bit-keyvaluesort.hpp [hunk 7]
```diff
@@ -0,0 +1,1276 @@
...
+        nan_count += _mm_popcnt_u32((int32_t)nanmask);
+        _mm512_mask_storeu_pd(arr, nanmask, ZMM_MAX_DOUBLE);
+        arr += 8;
+        arrsize -= 8;
+    }
+    return nan_count;
+}
+
+X86_SIMD_SORT_INLINE void
+replace_inf_with_nan_kv(double *arr, int64_t arrsize, int64_t nan_count)
```
Comment (r-devulap) [inline src/avx512-64bit-keyvaluesort.hpp:1237]: same as above. Move this to common header file and do not duplicate.

File: src/avx512-common-keyvaluesort.h [hunk 8]
```diff
@@ -0,0 +1,291 @@
...
+ */
+#define X86_SIMD_SORT_INLINE static __attribute__((always_inline))
+#define X86_SIMD_SORT_FINLINE static __attribute__((always_inline))
+#elif defined(__GNUC__)
+#define X86_SIMD_SORT_INLINE static inline
+#define X86_SIMD_SORT_FINLINE static __attribute__((always_inline))
+#else
+#define X86_SIMD_SORT_INLINE static
+#define X86_SIMD_SORT_FINLINE static
+#endif
```
Comment (r-devulap) [inline src/avx512-common-keyvaluesort.h:82]: lines 36-82 is duplicated. Move them to new header file and include that here and `avx512-common-qsort.h`

File: src/avx512-16bit-qsort.hpp [hunk 9]
```diff
@@ -686,7 +686,7 @@ replace_inf_with_nan(uint16_t *arr, int64_t arrsize, int64_t nan_count)
 }
 
 template <>
-void avx512_qsort(int16_t *arr, int64_t arrsize)
+inline void avx512_qsort(int16_t *arr, int64_t arrsize)
```
Comment (r-devulap) [inline src/avx512-16bit-qsort.hpp:689]: why `inline `these?

File: src/avx512-16bit-qsort.hpp [hunk 10]
```diff
@@ -695,7 +695,7 @@ void avx512_qsort(int16_t *arr, int64_t arrsize)
 }
 
 template <>
-void avx512_qsort(uint16_t *arr, int64_t arrsize)
+inline void avx512_qsort(uint16_t *arr, int64_t arrsize)
```
Comment (r-devulap) [inline src/avx512-16bit-qsort.hpp:698]: same. get rid of inline please.

File: src/avx512-32bit-qsort.hpp [hunk 11]
```diff
@@ -682,7 +682,7 @@ replace_inf_with_nan(float *arr, int64_t arrsize, int64_t nan_count)
 }
 
 template <>
-void avx512_qsort<int32_t>(int32_t *arr, int64_t arrsize)
+inline void avx512_qsort<int32_t>(int32_t *arr, int64_t arrsize)
```
Comment (r-devulap) [inline src/avx512-32bit-qsort.hpp:685]: remove inline.

File: src/avx512-32bit-qsort.hpp [hunk 12]
```diff
@@ -691,7 +691,7 @@ void avx512_qsort<int32_t>(int32_t *arr, int64_t arrsize)
 }
 
 template <>
-void avx512_qsort<uint32_t>(uint32_t *arr, int64_t arrsize)
+inline void avx512_qsort<uint32_t>(uint32_t *arr, int64_t arrsize)
```
Comment (r-devulap) [inline src/avx512-32bit-qsort.hpp:694]: ditto

File: src/avx512-32bit-qsort.hpp [hunk 13]
```diff
@@ -700,7 +700,7 @@ void avx512_qsort<uint32_t>(uint32_t *arr, int64_t arrsize)
 }
 
 template <>
-void avx512_qsort<float>(float *arr, int64_t arrsize)
+inline void avx512_qsort<float>(float *arr, int64_t arrsize)
```
Comment (r-devulap) [inline src/avx512-32bit-qsort.hpp:703]: ditto

File: src/avx512-64bit-keyvaluesort.hpp [hunk 14]
```diff
@@ -0,0 +1,1276 @@
...
+{
+    for (int64_t ii = arrsize - 1; nan_count > 0; --ii) {
+        arr[ii] = std::nan("1");
+        nan_count -= 1;
+    }
+}
+
+template <>
+inline void
+avx512_qsort_kv<int64_t>(int64_t *keys, uint64_t *indexes, int64_t arrsize)
```
Comment (r-devulap) [inline src/avx512-64bit-keyvaluesort.hpp:1247]: Are indices always expected to be a `uint64_t `?

**Review by r-devulap — CHANGES_REQUESTED** [review]
Still working on reviewing the code, meanwhile lets clean up all the duplicated code.

**Comment by ruclz** [comment]
> @ruclz Could you add benchmarking to know how fast this is relative to using std::sort? From issue #8:
> 
> > avx512_qsort_kv(double *keys, uint64_t *indexes, int64_t arrsize) is a little worse than my simple bubble sorting.

OK, I am coding for benchmarking.

**Comment by ruclz** [comment]
> @ruclz Could you add benchmarking to know how fast this is relative to using std::sort? From issue #8:
> 
> > avx512_qsort_kv(double *keys, uint64_t *indexes, int64_t arrsize) is a little worse than my simple bubble sorting.

image

The figure shows the performance of key-value bitonic sort and std::sort. There is no performance drop compared with std::sort.

File: src/avx512-64bit-keyvaluesort.hpp (hunk 5 above)
Comment (ruclz) [reply src/avx512-64bit-keyvaluesort.hpp:1194]: Because the arguments are two column pointer(keys and values), if we want to use std::sort, the input of std::sort is a struct, not two arguments, we need create a new two-dimension vector to store key and index, it need to malloc extra space.

File: benchmarks/bench.hpp [hunk 15]
```diff
@@ -72,3 +79,49 @@ std::tuple<uint64_t, uint64_t> bench_sort(const std::vector<T> arr,
...
+        values_bckup = values;
+    }
+    uint64_t avx_sort = std::accumulate(runtimes1.end() - lastfew,
+                                        runtimes1.end(),
+                                        (uint64_t)0)
+            / lastfew;
+
+    for (uint64_t ii = 0; ii < iters; ++ii) {
+        start = cycles_start();
+        std::sort(sortedaar_bckup.begin(),
```
Comment (r-devulap) [inline benchmarks/bench.hpp:113]: Not sure if std::sort is a fair comparison. The way the key-values are arranged in memory for `std::vector<sorted_t<K, V>>` is very different when compared two distinct vectors  `std::vector<K>, std::vector<V>`. I would think this can alter performance significantly because you can no longer load continuous chunk of memory into ZMM registers. Does Oceanbase have two separate vectors for key and values (I assume they do cos that's the way you implemented the sort). How does Oceanbase currently sort them?

File: benchmarks/bench-tgl.out [hunk 16]
```diff
@@ -1,28 +1,29 @@
...
-| uniform random  | double      | 100000     | 3596886         | 21849587  | 6.1      |
-| uniform random  | double      | 1000000    | 47724381        | 258014177 | 5.4      |
-| uniform random  | uint16_t    | 10000      | 84732           | 1548275   | 18.3     |
-| uniform random  | uint16_t    | 100000     | 1406417         | 19632858  | 14.0     |
-| uniform random  | uint16_t    | 1000000    | 17119960        | 214085305 | 12.5     |
-| uniform random  | int16_t     | 10000      | 84703           | 1547726   | 18.3     |
-| uniform random  | int16_t     | 100000     | 1442726         | 19705242  | 13.7     |
-| uniform random  | int16_t     | 1000000    | 20210224        | 212137465 | 10.5     |
-|-----------------+-------------+------------+-----------------+-----------+----------|
+| -----------------+-------------+------------+-----------------+-----------+---------- |
```
Comment (r-devulap) [inline benchmarks/bench-tgl.out:1]: please dont check in this file.

File: src/avx512-64bit-keyvaluesort.hpp [hunk 17]
```diff
@@ -0,0 +1,909 @@
...
+                                          left + 6 * size,
+                                          left + 7 * size,
+                                          left + 8 * size);
+    zmm_t key_vec = vtype::template i64gather<sizeof(type_t)>(rand_index, keys);
+
+    index_type index_vec;
+    zmm_t sort;
+    index_vec = zmm_vector<uint64_t>::template i64gather<sizeof(uint64_t)>(
+            rand_index, indexes);
+    sort = sort_zmm_64bit<vtype>(key_vec, index_vec);
```
Comment (r-devulap) [inline src/avx512-64bit-keyvaluesort.hpp:801]: pivot is picked just based on keys and not index, right? `index_vec` is never really used anywhere in this function, which means you do not need a new `get_pivot_64bit` function. You can move this function to `avx512-64bit-common.h` and just call that instead. You will also need to move `sort_zmm_64_bit<vtype>` into the common file. This should eliminate the need to calling gather function twice and save you some cpu cycles.

**Review by r-devulap — CHANGES_REQUESTED** [review]

File: benchmarks/bench.hpp (hunk 15 above)
Comment (xtangxtang) [reply benchmarks/bench.hpp:113]: ObStoreRow is the base unit to do the sort in Oceanbase. The key and value are all ObObject class type in the ObStoreRow. There is no sperate vector for key and value. The ObStoreRow is passed to the std::sort in Oceanbase. std::sort first get the key type from ObObject and then do the sort work.

Comment (xtangxtang) [reply benchmarks/bench.hpp:113]: To enable the AVX512 sort, we need extract the key from the ObStoreRow and then pass the key and ObStoreRow as the value to avx512_sort(key, value) like API. That's way we need the key-value interface

**Comment by ruclz** [comment]
We have moved the get_povit_64bit function and sort_zmm_64bit function to avx512-64bit-common.h, and modified the bench-tgl.out to origin file. Please help to review again, thanks.

**Comment by r-devulap** [comment]
@ruclz please fix the author name and get rid of the unnecessary struct and we should be good to go.

**Comment by ruclz** [comment]
Yeah, we have fixed the authors name in the last commit.

**Comment by r-devulap** [comment]
LGTM, Thanks @ruclz!

**Comment by r-devulap** [comment]
Did a little bit of code cleanup, see #15
